lib/spack/llnl/util/tty/log.py

```diff
@@ -1025,9 +1021,9 @@ def _writer_daemon(
         if isinstance(log_file, io.StringIO):
             control_pipe.send(log_file.getvalue())
         log_file_wrapper.close()
-        close_connection_and_file(read_multiprocess_fd, in_pipe)
-        if stdin_multiprocess_fd:
-            close_connection_and_file(stdin_multiprocess_fd, stdin)
+        in_pipe.close()
+        if stdin:
+            stdin.close()
```

Comment:
These 3 lines are the only ones that matter?

Reply:
This suggestion works on Linux OK, but not on Mac for me. E.g. if I undo the change in `MultiProcessFd`

```
diff --git a/lib/spack/llnl/util/tty/log.py b/lib/spack/llnl/util/tty/log.py
index 30f97a924d..b3b6bdb307 100644
--- a/lib/spack/llnl/util/tty/log.py
+++ b/lib/spack/llnl/util/tty/log.py
@@ -369,10 +369,9 @@ def fd(self):
         Connection object.
         """
         if self._connection:
-            self._fd = self._connection._handle
-            self._connection._handle = None
-            self._connection = None
-        return self._fd
+            return self._connection._handle
+        else:
+            return self._fd
 
     def close(self):
         """This only needs to be called if the user has not
```

I get:

```
Exception ignored in: <function _ConnectionBase.__del__ at 0x103998820>
Traceback (most recent call last):
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/multiprocessing/connection.py", line 137, in __del__
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/multiprocessing/connection.py", line 366, in _close
OSError: [Errno 9] Bad file descriptor
```

Did you test on both?

Reply:
In other words:

* If you just remove the `close()` operations here, you get (suppressed) warnings (which I think isn't great)
* If you just undo the changes to `MultiProcessFd`, you get (unsuppressed) warnings (on Mac OS, the original reason for this PR)

Reply:
Ah ok, I did not try on macOS. Will look again on Monday. Takes a bit of time to understand what is going on here...

Would still be nice to not have to set `._handle = None`. Suggests to me we're using it wrong.

FWIW, a debug build of Python (spack install python@... +debug) gives the same warnings as Python 3.13 and more, which is what I tried on Linux earlier today. Will try the same on macOS later.